### ISSUE TITLE
Console / Trace / Debugger capture support

### ISSUE THREAD

**Issue by jarrettv** [xunit/xunit#1730]
Sometimes you use tests as a way to develop code that is hosted by another runtime. For example an AWS lambda function. You can no longer easily use xunit(since 2.0) because you can't see output as you step through code.

**Comment by bradwilson** [comment]
Related: #892

**Comment by sillyousu** [comment]
any update?

**Comment by bradwilson** [comment]
Work has not yet begun on 3.0. It will be quite a while before it's ready.

**Comment by bradwilson** [comment]
Some existing code we may be able to borrow/take inspiration from: https://github.com/SimonCropp/XunitLogger

**Comment by SimonCropp** [comment]
@bradwilson would be more than happy to donate that code to xunit. or help with a PR

**Comment by bradwilson** [comment]
The intention is to implement all of this capture in terms of output that gets funneled into `ITestOutputHelper`. In order to make this work, I'm going to expand `ITestOutputHelper` to offer `Write` methods in addition to `WriteLine` methods.

This will not change the contract of `ITestOutput`, which is that it is only sent when a full line of output has been written. This means we'll be breaking apart lines based on [`Environment.NewLine`](https://learn.microsoft.com/dotnet/api/system.environment.newline) and sending only when full lines are available. If [live output is enabled](https://xunit.net/docs/configuration-files#showLiveOutput), you will only see that output when line breaks are encountered.

Since all output is funneled through `ITestOutputHelper`, that means that any output from any captured source will be interleaved with all the others. A downside of this is that there will be no way to identify which source a captured piece of output comes from. An upside of this is that the output should represent a linear time perspective on output. _If you need to explicitly see the output from multiple sources independently (for example, seeing StdOut and StdErr separately), you should capture those yourself in the test, and attach them to the test results using `TestContext.Current.AddAttachment`._

### Implementation detail

Access to `ITestOutputHelper` will come via `TestContext.Current`. The context is what understands the currently executing test and will allow us to associate your output with the right test.

**Important note:** The `TestContext` instance is propagated with [`AsyncLocal`](https://learn.microsoft.com/dotnet/api/system.threading.asynclocal-1). Any attempt to access the `TestContext` that take place outside the context of a test method's [`ExecutionContext`](https://learn.microsoft.com/dotnet/api/system.threading.executioncontext) will end silently discarding the output, as it's the execution context that makes `AsyncLocal` work. The most common way you will end up leaving the execution context is by awaiting a task result which was appended with `.ConfigureAwait(false)` (that is what `.ConfigureAwait(false)` does). A less common way will be any background thread that the test (or code under test) creates which doesn't propagate the `ExecutionContext` (via `.Capture` and `.Restore`).

Short answer: don't use `.ConfigureAwait(false)` in your unit tests. We have an [analyzer which catches this behavior](https://xunit.net/xunit.analyzers/rules/xUnit1030), but only in the context of test methods.

**Comment by bradwilson** [comment]
The implementation will be via two attributes: `[CaptureConsole]` and `[CaptureTrace]`.

`[CaptureConsole]` will allow the user to specify whether they want to capture `Console.Out` and/or `Console.Error` via settable properties (`CaptureOut` and `CaptureError`), and it will capture both by default. Note that capturing `Console.Out` also captures the direct writing functions `Console.Write` and `Console.WriteLine` as they are functionally identical to `Console.Out.Write` and `Console.Out.WriteLine`.

I had intended to have `[CaptureTrace]` offer the same options for capturing `Debug` vs. `Trace`, but as it turns out, in .NET Core they combined the listeners into a single list, so you either capture both or neither. This is different from .NET Framework, which kept separate listener lists. In order to get consistent behavior on both platforms, therefore, `[CaptureTrace]` will always capture both.

Given that these attributes are affecting a system-wide shared resource (`Console.Out`, `Console.Error`, `Trace.Listeners`, etc.) these will be implemented as an attribute that is only legal at the assembly level. Here is an example acceptance test:

```csharp
[assembly: CaptureConsole]

public class TestClass
{
    [Fact]
    public static void AcceptanceTest()
    {
        Console.WriteLine("This is a line of text from Console.WriteLine");
        TestContext.Current.TestOutputHelper?.WriteLine("This is a line of text from the output helper");
        Console.Error.WriteLine("This is a line of text from Console.Error.WriteLine");

        Assert.Equal(
            "This is a line of text from Console.WriteLine" + Environment.NewLine +
            "This is a line of text from the output helper" + Environment.NewLine +
            "This is a line of text from Console.Error.WriteLine" + Environment.NewLine,
            TestContext.Current.TestOutputHelper?.Output
        );
    }
}
```

As mentioned above, all of these offer both `Write` and `WriteLine` variants now, so you can write partial output now (a feature that was requested a few times for `ITestOutputHelper` but I had not implemented until now).

**Comment by bradwilson** [comment]
Available in `0.3.0-pre.28` https://xunit.net/docs/using-ci-builds
